### PR TITLE
ostest:priority_inheritance Added priority restoration test

### PR DESCRIPTION
**Pull request by davids5**
## Summary

See https://github.com/apache/incubator-nuttx/issues/6310

## Impact

Loss of system priorities

## Testing

master nuttx

### CODE REVIEW

File: testing/ostest/prioinherit.c [hunk 1]
```diff
@@ -588,6 +665,61 @@ void priority_inheritance(void)
...
+  for (i = 0; i < NUMBER_OF_COMPETING_THREADS; i++)
+    {
+      g_priority_tracking[i] = 0;
+      snprintf(name, sizeof(name), "Task%1d", i);
+      snprintf(args[0], sizeof(args[0]), "%d", i);
+      snprintf(args[1], sizeof(args[1]), "%d", i * 10000);
+      snprintf(args[2], sizeof(args[2]), "%d", i == 0 ? 100000 : 1000);
+
+      pids[i] = task_create(name, priority, 1024, adversary,
+                            (FAR char * const *) argv);
```
Comment (xiaoxiang781216) [inline testing/ostest/prioinherit.c:688]: remove space before args

File: testing/ostest/prioinherit.c [hunk 2]
```diff
@@ -416,6 +487,12 @@ void priority_inheritance(void)
   int my_pri;
   int status;
   int i;
+  char pids[NUMBER_OF_COMPETING_THREADS];
+  char args[3][32];
+  char *argv[4];
```
Comment (xiaoxiang781216) [inline testing/ostest/prioinherit.c:492]: add FAR

File: testing/ostest/prioinherit.c [hunk 3]
```diff
@@ -88,10 +93,76 @@ static sem_t g_sem;
...
+      us--;
+    }
+  while (us > 0);
+}
+
+/****************************************************************************
+ *
+ ****************************************************************************/
+
+static int adversary(int argc, char *argv[])
```
Comment (xiaoxiang781216) [inline testing/ostest/prioinherit.c:149]: add FAR

File: testing/ostest/prioinherit.c [hunk 4]
```diff
@@ -588,6 +667,61 @@ void priority_inheritance(void)
...
+
+  if (restoration_result != 0)
+    {
+      printf("priority_inheritance: ERROR: FAIL Priorities were not "
+             "correctly restored.\n");
+    }
+  else
+    {
+      printf("priority_inheritance: PASSED Priority were correctly"
+          " restored.\n");
```
Comment (pkarashchenko) [inline testing/ostest/prioinherit.c:722]:
```suggestion
      printf("priority_inheritance: PASSED Priority were correctly"
             " restored.\n");
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

**Comment by masayuki2009** [comment]
I noticed that `qemu-intel64:ostest` causes a deadlock.

```
SeaBIOS (version rel-1.12.0-59-gc9ba5276e321-prebuilt.qemu.org)


iPXE (http://ipxe.org) 00:03.0 CA00 PCI2.10 PnP PMM+0FF907F0+0FEF07F0 CA00
                                                                               


Booting from Hard Disk...
Boot failed: could not read the boot disk

Booting from Floppy...
Boot failed: could not read the boot disk

Booting from DVD/CD...
stdio_test: write fd=1
stdio_test: Standard I/O Check: printf
stdio_test: write fd=2
ostest_main: putenv(Variable1=BadValue3)
ostest_main: setenv(Variable1, GoodValue1, TRUE)
ostest_main: setenv(Variable2, BadValue1, FALSE)
ostest_main: setenv(Variable2, GoodValue2, TRUE)
ostest_main: setenv(Variable3, Variable3, FALSE)
ostest_main: setenv(Variable3, Variable3, FALSE)
show_variable: Variable=Variable1 has value=GoodValue1
show_variable: Variable=Variable2 has value=GoodValue2
show_variable: Variable=Variable3 has value=GoodValue3
ostest_main: Started user_main at PID=2

...

user_main: priority inheritance test
priority_inheritance: Started
priority_inheritance: Starting lowpri_thread-1 (of 3) at 1
priority_inheritance: Set lowpri_thread-1 priority to 1
priority_inheritance: Starting lowpri_thread-2 (of 3) at 1
priority_inheritance: Set lowpri_thread-2 priority to 1
priority_inheritance: Starting lowpri_thread-3 (of 3) at 1
priority_inheritance: Set lowpri_thread-3 priority to 1
priority_inheritance: Waiting...
lowpri_thread-1: Started
lowpri_thread-1: initial priority: 1
lowpri_thread-1: Waiting for the midle pri task to run
    g_middlestate:  0
    g_highstate[0]: 0
    g_highstate[1]: 0
    g_highstate[2]: 0
    I still have a count on the semaphore
lowpri_thread-2: Started
lowpri_thread-2: initial priority: 1
lowpri_thread-2: Waiting for the midle pri task to run
    g_middlestate:  0
    g_highstate[0]: 0
    g_highstate[1]: 0
    g_highstate[2]: 0
    I still have a count on the semaphore
lowpri_thread-3: Started
lowpri_thread-3: initial priority: 1
lowpri_thread-3: Waiting for the midle pri task to run
    g_middlestate:  0
    g_highstate[0]: 0
    g_highstate[1]: 0
    g_highstate[2]: 0
    I still have a count on the semaphore
lowpri_thread-1: Waiting for the midle pri task to run
    g_middlestate:  0
    g_highstate[0]: 0
    g_highstate[1]: 0
    g_highstate[2]: 0
    I still have a count on the semaphore
lowpri_thread-2: Waiting for the midle pri task to run
    g_middlestate:  0
    g_highstate[0]: 0
    g_highstate[1]: 0
    g_highstate[2]: 0
    I still have a count on the semaphore
lowpri_thread-3: Waiting for the midle pri task to run
    g_middlestate:  0
    g_highstate[0]: 0
    g_highstate[1]: 0
    g_highstate[2]: 0
    I still have a count on the semaphore
priority_inheritance: Starting medpri_thread at 99
priority_inheritance: Set medpri_thread priority to 99
priority_inheritance: Waiting...
medpri_thread: Started ... I won't let go of the CPU!
priority_inheritance: Starting highpri_thread-1 (of 3) at 255
priority_inheritance: Set highpri_thread-1 priority to 255
highpri_thread-1: Started
priority_inheritance: Starting highpri_thread-2 (of 3) at 255
priority_inheritance: Set highpri_thread-2 priority to 254
highpri_thread-2: Started
priority_inheritance: Starting highpri_thread-3 (of 3) at 255
priority_inheritance: Set highpri_thread-3 priority to 253
highpri_thread-3: Started
priority_inheritance: Waiting for highpri_thread-1 to complete
highpri_thread-1: Calling sem_wait()
lowpri_thread-3: Sem count: -1, No. highpri thread: 1
highpri_thread-1: SUCCESS midpri_thread is still running!
highpri_thread-1: Okay... I'm done!
priority_inheritance: Waiting for highpri_thread-2 to complete
highpri_thread-2: Calling sem_wait()
highpri_thread-2: SUCCESS midpri_thread is still running!
highpri_thread-2: Okay... I'm done!
priority_inheritance: Waiting for highpri_thread-3 to complete
highpri_thread-3: Calling sem_wait()
highpri_thread-3: SUCCESS midpri_thread is still running!
highpri_thread-3: Okay... I'm done!
priority_inheritance: Waiting for medpri_thread to complete
medpri_thread: Okay... I'm done!
priority_inheritance: Waiting for lowpri_thread-1 to complete
lowpri_thread-2: Sem count: 1, No. highpri thread: 0
lowpri_thread-2: SUCCESS the middle priority task has already exitted!
               g_middlestate:  3 sem count=1
               g_highstate[0]: 3
               g_highstate[1]: 3
               g_highstate[2]: 3
lowpri_thread-2: SUCCESS priority before sem_post: 1
lowpri_thread-2: SUCCESS final priority: 1
lowpri_thread-2: Okay... I'm done!
lowpri_thread-1: Sem count: 2, No. highpri thread: 0
lowpri_thread-1: SUCCESS the middle priority task has already exitted!
               g_middlestate:  3 sem count=2
               g_highstate[0]: 3
               g_highstate[1]: 3
               g_highstate[2]: 3
lowpri_thread-1: SUCCESS priority before sem_post: 1
lowpri_thread-1: SUCCESS final priority: 1
lowpri_thread-1: Okay... I'm done!
priority_inheritance: Waiting for lowpri_thread-2 to complete
priority_inheritance: Waiting for lowpri_thread-3 to complete
lowpri_thread-3: SUCCESS priority before sem_post: 255
lowpri_thread-3: SUCCESS final priority: 1
lowpri_thread-3: Okay... I'm done!
priority_inheritance: Restoration Test:

```

**Pull request merged**